src/hosted-buttons/utils.js

```diff
@@ -126,16 +130,37 @@ export const buildHostedButtonCreateOrder = ({
           merchant_id: merchantId,
           ...userInputs,
         }),
-      }).then(({ body }) => {
-        return body.context_id;
-      });
+      })
+        .then(({ body }) => body.context_id || onError(body.name))
+        .catch(() => onError("REQUEST_FAILED"));
     });
   };
 };
 
+export const buildOpenPopup: BuildOpenPopup = ({ popupFallback, selector }) => {
+  return (url) => {
+    try {
+      popup(url, {
+        width: DEFAULT_POPUP_SIZE.WIDTH,
+        height: DEFAULT_POPUP_SIZE.HEIGHT,
+      });
+    } catch (e) {
+      const div = document.createElement("div");
+      div.classList.add(popupFallbackClassName);
+      div.innerHTML = popupFallback.replace("#", url);
+      const container =
+        typeof selector === "string"
+          ? document.querySelector(selector)
+          : selector;
+      container?.appendChild(div);
+    }
+  };
+};
```

Comment:
I'm not sure I like this implementation. I did it so that I didn't have to pass `popupFallback` and `selector` down into `onApprove`. e.g. the alternative to this would be:

```js
onApprove: buildHostedButtonOnApprove({
  merchantId,
  hostedButtonId,
  selector,
  popupFallback
})
```

and then inside of `buildHostedButtonOnApprove` would be the contents of this function. I did it this way so that I could easily unit-test it without managing passing around props but now that I'm looking at it in the GitHub diff it looks a little confusing. wdyt?

Reply:
Ease of testing is a good reason in my book for going this route. Thanks for explaining this!